server/src/internalClusterTest/java/org/elasticsearch/repositories/blobstore/BlobStoreCorruptionIT.java

```diff
@@ -65,17 +68,49 @@ public void testCorruptionDetection() throws Exception {
 
         // detect corruption by taking another snapshot
         if (corruptedFileType == RepositoryFileType.SHARD_GENERATION) {
-            corruptionDetectors.add(exceptionListener -> {
-                logger.info("--> taking another snapshot");
-                client().admin()
-                    .cluster()
-                    .prepareCreateSnapshot(TEST_REQUEST_TIMEOUT, repositoryName, randomIdentifier())
-                    .setWaitForCompletion(true)
-                    .execute(exceptionListener.map(createSnapshotResponse -> {
-                        assertNotEquals(SnapshotState.SUCCESS, createSnapshotResponse.getSnapshotInfo().state());
-                        return new ElasticsearchException("create-snapshot failed as expected");
-                    }));
-            });
+            if (Files.exists(corruptedFile)) {
+                corruptionDetectors.add(exceptionListener -> {
+                    logger.info("--> taking another snapshot");
+                    client().admin()
+                        .cluster()
+                        .prepareCreateSnapshot(TEST_REQUEST_TIMEOUT, repositoryName, randomIdentifier())
+                        .setWaitForCompletion(true)
+                        .execute(exceptionListener.map(createSnapshotResponse -> {
+                            assertNotEquals(SnapshotState.SUCCESS, createSnapshotResponse.getSnapshotInfo().state());
+                            return new ElasticsearchException("create-snapshot failed as expected");
+                        }));
+                });
+            } else {
```

Comment:
I wonder if validating following is relevant:

1. That we can restore without the missing file?
2. That taking a new snapshot that includes changes to the shard repairs the situation (I think it would?) for new snapshots, i.e., we no longer need the fallback reading?

Reply:
Yeah actually we already have a test for the behaviour when this file is corrupt, see `testSnapshotWithCorruptedShardIndexFile`. So in c13297f70113763c83e9e4e8bccf8a764fe6caac I reverted back to having this test delete the file again, and made it try more interesting combinations of snapshot creations and restores to make sure it still works as we expect.

Reply:
IIUC, both cloning and deleting snapshot should also write a new shard generation file and fix the issue. Should we randomly test for them as well?

Reply:
Curious now why this works for restore - but I suppose we somehow avoid reading that file?

Reply:
> both cloning and deleting snapshot should also write a new shard generation file and fix the issue

++ see b6434df7c220099f4e383ddea1e5b7ff850544db